### ISSUE TITLE
Unable to load categorical data from .arff files using LoadARFF

### ISSUE THREAD

**Issue by pesoke** [mlpack/mlpack#2478]
My code:
arma::mat data;
data::DatasetInfo info;
data::LoadARFF("data.csv", data, info);

At run time, this error is present: 
terminate called after throwing an instance of 'std::runtime_error'
  what():  Parse error at line 38 token 8: "B".
Abort (core dumped)

I tried again same file but without the categorical data and it works well.

**Comment by rcurtin** [comment]
Can you provide the contents of `data.csv`? (Also shouldn't it be `data.arff`?)

**Comment by pesoke** [comment]
This is a portion of it:

@relation analcatdata-braziltourism

@attribute 'Age' INTEGER
@attribute 'Sex' {0,1,4}
@attribute 'Income' REAL
@attribute 'Travel_cost' REAL
@attribute 'Access_road' {0,1}
@attribute 'Active' {0,1,2,3,4,5,6}
@attribute 'Passive' {0,1,2,3,4}
@attribute 'Logged_income' REAL
@attribute 'Categorical Data' { A, B, C}

@data
41,0,2187.5,186.9,0,0,3,7.690514618,B
41,1,406.25,48.0003,0,1,3,6.006968734,B
42,0,2031.25,48.0003,0,2,1,7.616406646,B
24,1,1562.5,161.7,0,2,3,7.354042382,B
32,1,781.25,48.0003,0,0,0,6.660895201,B


Also, you're right, it's data.arff not .csv

**Comment by rcurtin** [comment]
@pesoke thanks for the info.  I did some digging and ended up opening #2486 to fix the issue.  If you like, would you mind trying that branch and seeing if it solves your problem?  I think it should.  Thanks! :+1:

**Comment by pesoke** [comment]
Hi, the method works now for categorical data but the mapping is not consistent. For example I tried on another file which used 18 letters as labels and they were mapped from 18 - 35, instead of 0 to 17. The following is a section of the arff file mentioned:
```@attribute "ACTIVITY" { A, B, C, D, E, F, G, H, I, J, K, L, M, O, P, Q, R, S }```

This meant that when passed through the following FFN, I had to first print out the mapping, then either preprocess the data using   ```for (size_t i = 0; i < trainLabelsTemp.n_cols; ++i)    trainLabels(i) = trainLabelsTemp(i) - 17; ``` or  change the final layer's output to 35 instead of 18 which is the number of classes. Else I will get the following error ```error: Mat::operator(): index out of boundsterminate called after throwing an instance of 'std::logic_error'what():  Mat::operator(): index out of bounds``` as the label is expected to be within the range of [1, num_classes]

```  FFN<> model;
  model.Add<Linear<>>(trainData.n_rows, 100);
  model.Add<SigmoidLayer<>>();
  model.Add<Linear<>>(100, 18);
  model.Add<LogSoftMax<>>();```

**Comment by rcurtin** [comment]
Oops, @pesoke, sorry I didn't get it the first time.  I'll try and reproduce this and fix it shortly. :+1:

**Comment by rcurtin** [comment]
Hi @pesoke, I just opened #2516 which should hopefully fix your issue.  I tested the updated ARFF loader with the WISDM smartphone data (which is I think where your ARFF file came from):
https://archive.ics.uci.edu/ml/datasets/WISDM+Smartphone+and+Smartwatch+Activity+and+Biometrics+Dataset+

Let me know if that still doesn't fix your issue and I can try again. :smile:

**Comment by rcurtin** [comment]
I'll go ahead and close this since #2516 is merged, but please, if there are further issues just post here and we can reopen the issue, or alternately you can open a new issue.  Up to you.  (In any case I hope that isn't necessary and that your issue is fixed :smile:)

**Comment by pesoke** [comment]
I tried it out and now  {A, B, C, D, E, F, G, H, I, J, K, L, M, O, P, Q, R, S } is mapped to 0 to 17 which is better than 18 to 35. However, the same error as before is still thrown as the model.Train expects the labels to be between [1, 18]. Is there a way for it to map from 1 to 18 instead of 0 to 17 so we do not have to the conversion on our own?

**Comment by rcurtin** [comment]
The `NegativeLogLikelihood<>` loss function is a bit unique in mlpack in that it expects labels starting from 1---but basically everything else in mlpack expects it starting from 0.  I'm going to open a patch to change that (although it may require a bit of discussion and changes since it's a pretty fundamental part of the neural network code), but for now, I think the workaround that I would suggest would be to simply do `data.col(i) += 1` where `data` is your matrix, and `i` is the column that got mapped from 0 to 17.  I'll tag this issue when that PR is open.

**Comment by pesoke** [comment]
I see. I will use the workaround for now. Thank you.

**Comment by rcurtin** [comment]
I opened #2534 to handle the issue; it may be a little while until it's merged (and it will have to be a part of a 4.0.0 release since it's not reverse-compatible).  Feel free to cherry-pick that commit into your work if you like, or wait to see what happens with the PR. :+1:
